src/utils/manager.py

```diff
@@ -185,55 +186,128 @@ def mount(
         endpoint, port = _translate(endpoint)
     else:
         port = None
 
-    if not (target := pathlib.Path(mountpoint)).exists():
-        _logger.debug(f"Creating mountpoint {target}")
-        target.mkdir()
+    # Avoids TOCTOU
+    _logger.debug(f"Ensuring mountpoint at {mountpoint} exists")
+    pathlib.Path(mountpoint).mkdir(exist_ok=True)
+
+    autofs_id = _mountpoint_to_autofs_id(mountpoint)
 
-    cmd = ["mount", "-t", "nfs"]
+    mount_opts = ""
     if options:
         if port:
             options.append(f"port={port}")
-        cmd.extend(["-o", ",".join(options)])
-    cmd.extend([endpoint, target])
+        mount_opts = ",".join(options)
+
+    _logger.debug(f"Mounting NFS share endpoint {endpoint} at {mountpoint} with options {options}")
+
+    master = f"/- /etc/auto.{autofs_id}" + (f" {mount_opts}" if mount_opts else "")
+
+    write_file(f"/etc/auto.master.d/{autofs_id}.autofs", master)
+    write_file(f"/etc/auto.{autofs_id}", f"{mountpoint} {endpoint}")
+
     try:
-        _logger.debug(f"Mounting NFS share endpoint {endpoint} at {target} with options {options}")
-        subprocess.run(cmd, stderr=subprocess.PIPE, check=True, text=True)
-    except subprocess.CalledProcessError as e:
-        _logger.error(f"{e} Reason:\n{e.stderr}")
-        if "Operation not permitted" in e.stderr and not supported():
+        systemd.service_reload("autofs", restart_on_failure=True)
+    except systemd.SystemdError as e:
+        _logger.error(f"Failed to mount {endpoint} at {mountpoint}. Reason:\n{e}")
+        if "Operation not permitted" in str(e) and not supported():
             raise Error("Mounting NFS shares not supported on LXD containers")
-        raise Error(f"Failed to mount {endpoint} at {target}")
+        raise Error(f"Failed to mount {endpoint} at {mountpoint}")
 
 
-def umount(mountpoint: Union[str, os.PathLike], force: bool = False) -> None:
+def umount(mountpoint: Union[str, os.PathLike]) -> None:
     """Unmount an NFS share.
 
     Args:
         mountpoint: NFS share mountpoint to unmount.
-        force: Pass `--force` option to umount. Default: False.
 
     Raises:
         Error: Raised if NFS share umount operation fails.
     """
-    cmd = ["umount"]
-    if force:
-        cmd.append("--force")
-    cmd.append(mountpoint)
+
+    def unlink_file(path: Union[str, os.PathLike]) -> None:
+        """Deletes the file in `path` if it exists.
+
+        Raises:
+            Error: Raised if the file could not be deleted.
+        """
+        try:
+            pathlib.Path(path).unlink(missing_ok=True)
+        except OSError as e:
+            _logger.error(f"Failed to delete file {path}. Reason:\n{e}")
+            raise Error(f"Failed to unmount {mountpoint}")
+
+    _logger.debug(f"Unmounting NFS share at mountpoint {mountpoint}")
+    autofs_id = _mountpoint_to_autofs_id(mountpoint)
+
+    unlink_file(f"/etc/auto.{autofs_id}")
+    unlink_file(f"/etc/auto.master.d/{autofs_id}.autofs")
```

Comment:
Why are you creating a nested function here for unlinking the file? It is difficult to unit test this nested function since we cannot call or patch it directly. I'd prefer we used `.unlink` directly and did not wrap it with a nested function:

 ```suggestion
    _logger.debug(f"Unmounting NFS share at mountpoint {mountpoint}")
    autofs_id = _mountpoint_to_autofs_id(mountpoint)
    pathlib.Path(f"/etc/auto.{autofs_id}").unlink(missing_ok=True)    
    pathllib.Path(f"/etc/auto.master.d/{autofs_id}.autofs").unlink(missing_ok=True)
```

Also, just a style preference, but since unlinking the file and looking up the autofs id a related, I removed the newline between the two. Typically I group statements together if they are related as it helps the code flow better :sweat_smile:

Reply:
Talked about this in the standup and decided to just go with forwarding the error to the admin as it is. We can return more specific errors in the future if there's a need for it.